File: data-structures/non-linear/binary-tree/iterative/BinaryTree.cpp
#include <iostream>
#include <queue>
#include <stack>
using namespace std;

class Node {
public:
  int data;
  Node *left;
  Node *right;

  Node(int value) {
    data = value;
    left = right = nullptr;
  }
};

void preorderTraversal(Node *root) {
  if (root == nullptr) {
    return;
  }

  stack<Node *> s;

  s.push(root);

  while (!s.empty()) {
    Node *node = s.top();

    s.pop();
    cout << node->data << " ";

    if (node->right) {
      s.push(node->right);
    }

    if (node->left) {
      s.push(node->left);
    }
  }
}

void inorderTraversal(Node *root) {
  stack<Node *> s;

  Node *node = root;

  while (node != nullptr || !s.empty()) {
    if (node != nullptr) {
      s.push(node);
      node = node->left;
    } else {
      node = s.top();
      s.pop();
      cout << node->data << " ";
      node = node->right;
    }
  }
}

void postorderTraversal(Node *root) {
  stack<Node *> s;

  Node *lastVisited = nullptr;
  Node *node = root;

  while (node != nullptr || !s.empty()) {
    if (node != nullptr) {
      s.push(node);
      node = node->left;
    } else {
      Node *peek = s.top();

      if (peek->right && lastVisited != peek->right) {
        node = peek->right;
      } else {
        cout << peek->data << " ";
        lastVisited = peek;
        s.pop();
      }
    }
  }
}

void levelorderTraversal(Node *root) {
  if (root == nullptr) {
    return;
  }

  queue<Node *> q;

  q.push(root);

  while (!q.empty()) {
    Node *node = q.front();

    q.pop();
    cout << node->data << " ";

    if (node->left) {
      q.push(node->left);
    }

    if (node->right) {
      q.push(node->right);
    }
  }
}

File: data-structures/non-linear/binary-tree/iterative/BinaryTree.ts
import { TreeNode } from "../../node/TreeNode";

function preOrderTraversal(root: TreeNode | null): void {
  if (root == null) return;

  let stack: Array<TreeNode> = new Array();
  let tempNode: TreeNode;
  stack.push(root);

  while (stack.length != 0) {
    //@ts-expect-error empty array pop returns undefined but that is already checked in the while loop
    tempNode = stack.pop();

    console.log(`${tempNode.getValue()}`);
    if (tempNode.getRight() != null) {
      //@ts-ignore checking has been done for right being null
      stack.push(tempNode.getRight());
    }

    if (tempNode.getLeft() != null) {
      //@ts-ignore
      stack.push(tempNode.getRight());
    }
  }
}

function inOrderTraversal(root: TreeNode | null): void {
  if (root == null) return;

  let stack: Array<TreeNode> = new Array();
  let node: TreeNode | null = root;

  while (node != null || stack.length != 0) {
    if (node != null) {
      stack.push(node);
      node = node.getLeft();
    } else {
      //@ts-expect-error
      node = stack.pop();
      console.log(`${node?.getValue()}`);
      //@ts-expect-error
      node = node.getRight();
    }
  }
}

function postOrderTraversal(root: TreeNode | null): void {
  let stack: Array<TreeNode> = new Array();

  let lastVisitedNode: TreeNode | null = null;
  let node: TreeNode | null = root;

  while (node != null || stack.length != 0) {
    if (node != null) {
      stack.push(node);
      node = node.getLeft();
    } else {
      let peek: TreeNode | null = stack[-1];

      if (peek.getRight() != null && lastVisitedNode != peek.getRight()) {
        node = peek.getRight();
      } else {
        console.log(`${peek.getValue()}`);
        lastVisitedNode = peek;
        stack.pop();
      }
    }
  }
}
